performance/admin.py

```diff
@@ -0,0 +1,447 @@
+from django.contrib import admin
+from io import StringIO
+from performance import models
+from django_q.tasks import async_task
+
+from frictionless import validate
+
+import os
+import csv
+import budgetportal
+
+VALID_REPORT_TYPES = ['Provincial Institutions Oversight Performance  Report',
+                      'National Institutions Oversight Performance  Report']
+
+
+def generate_import_report(report_type_validated, frictionless_report, not_matching_departments):
+    report = ""
+    if not report_type_validated:
+        report += "Report type must be for one of " + os.linesep
+        for report_type in VALID_REPORT_TYPES:
+            report += f"* {report_type} {os.linesep}"
+
+    if frictionless_report:
+        if not frictionless_report.valid:
+            for error in frictionless_report.tasks[0].errors:
+                report += f"* {error.message} {os.linesep}"
+
+    if len(not_matching_departments) > 0:
+        report += "Department names that could not be matched on import : " + os.linesep
+        for department in not_matching_departments:
+            report += f"* {department} {os.linesep}"
+
+    return report
+
+
+def save_imported_indicators(obj_id):
+    # read file
+    obj_to_update = models.EQPRSFileUpload.objects.get(id=obj_id)
+    full_text = obj_to_update.file.read().decode('utf-8')
+
+    # validate report type
+    report_type_validated = validate_report_type(full_text, obj_id)
+    if not report_type_validated:
+        return
+
+    # clean the csv & extract data
+    financial_year = get_financial_year(full_text)
+    sphere = get_sphere(full_text)
+    clean_text = full_text.split('\n', 3)[3]
+    f = StringIO(clean_text)
+    reader = csv.DictReader(f)
+    parsed_data = list(reader)
+
+    # find the objects
+    report_departments = set([x['Institution'] for x in parsed_data])
+    report_programmes = set([x['Programme'] for x in parsed_data])
+    num_imported = 0
+    total_record_count = len(parsed_data)
+    not_matching_departments = []
+    fy_obj = budgetportal.models.FinancialYear.objects.filter(slug=financial_year).first()
+    sphere_obj = budgetportal.models.Sphere.objects.filter(name=sphere, financial_year=fy_obj).first()
+
+    # clear department indicators
+    for department in report_departments:
+        for programme in report_programmes:
+            government_obj = budgetportal.models.Government.objects.filter(name=programme, sphere=sphere_obj).first()
+
+            department_obj = models.Department.objects.filter(name=department, government=government_obj).first()
+            models.Indicator.objects.filter(department=department_obj).delete()
```

Comment:
This works, I think, but it holds some assumptions about how the data is selected and included in the file. Which I at least want to highlight. I think it's ok at the moment - you select "Eastern Cape" and "Western Cape" and "Education" and "Health" and I think you will then necessarily have data for all those. But if the data selection interface worked differently, you might have deleted more than you had replacement data for.

If the data selection worked differently, e.g. if you could select a department with a given name in one province but not another, you could have rows 

```
Health | Eastern Cape
Education | Western Cape
```

and you end up deleting

```
(Health, Eastern Cape)
(Health, Western Cape)
(Education, Eastern Cape)
(Education, Western Cape)
```

Reply:
so there's nothing to change here yet, but I got scared for a minute.

Reply:
my understanding was that when you upload a csv file, you basically want to reset the data for the departments you have in your file. so I thought it would be ok to delete `(Health, Western Cape)` data even if your file just contains `(Health, Eastern Cape)`